### DISCUSSION TITLE
from_disk partially broken in spacy-nightly

### DISCUSSION THREAD

**Discussion by nopper** [explosion/spaCy#6296] [Help: Coding & Implementations] [answered]
I am using spacy-nightly and I am trying to copy a textcat pipe from one model to another. The pseudo-code looks like this:

```
nlp = spacy.load("new-model")
textcat = nlp.add_pipe("textcat", name="textcat-whatever")
textcat.from_disk("prev-model/textcat")
```

But I am hitting an exception because the vocab is not present in prev-model's textcat module. Here's the stack trace:

```
  File "scripts/test.py", line 18, in main
    textcat.from_disk(textcat_path.absolute())
  File "spacy/pipeline/trainable_pipe.pyx", line 331, in spacy.pipeline.trainable_pipe.TrainablePipe.from_disk
  File "<REDACTED>/lib/python3.8/site-packages/spacy/util.py", line 1085, in from_disk
    reader(path / key)
  File "spacy/pipeline/trainable_pipe.pyx", line 329, in spacy.pipeline.trainable_pipe.TrainablePipe.from_disk.lambda8
  File "spacy/vocab.pyx", line 466, in spacy.vocab.Vocab.from_disk
  File "spacy/strings.pyx", line 238, in spacy.strings.StringStore.from_disk
  File "<REDACTED>//lib/python3.8/site-packages/srsly/_json_api.py", line 48, in read_json
    file_path = force_path(path)
  File "<REDACTED>//lib/python3.8/site-packages/srsly/util.py", line 24, in force_path
    raise ValueError(f"Can't read file: {location}")
ValueError: Can't read file: <REDACTED>/prev-model/textcat/vocab/strings.json
```

I guess the issue is around these lines:

https://github.com/explosion/spaCy/blob/270c836bd6b23eb803694d46add6d92989e26f5b/spacy/pipeline/trainable_pipe.pyx#L329
https://github.com/explosion/spaCy/blob/270c836bd6b23eb803694d46add6d92989e26f5b/spacy/pipeline/trainable_pipe.pyx#L305
https://github.com/explosion/spaCy/blob/270c836bd6b23eb803694d46add6d92989e26f5b/spacy/pipeline/trainable_pipe.pyx#L288
https://github.com/explosion/spaCy/blob/270c836bd6b23eb803694d46add6d92989e26f5b/spacy/pipeline/trainable_pipe.pyx#L265

which should instead be guarded by an if, eg:

```
        if hasattr(self, "vocab") and self.vocab is not None:
            deserialize["vocab"] = lambda p: self.vocab.from_disk(p)
```

**Comment by adrianeboyd** [comment]
This requires a slightly different workflow in v3. Use the `source` argument with `nlp.add_pipe` to load a particular component from an existing model: https://nightly.spacy.io/api/language#add_pipe

```python
nlp.add_pipe("textcat", source=spacy.load("prev-model"))
```

Edited to add:

There are also some examples in the v3 migration guide: https://nightly.spacy.io/usage/v3#migrating-add-pipe

**Comment by nopper** [comment]
Thanks @adrianeboyd for the prompt reply. You suggestions indeed solved the issue on my side.

I guess, though, that the `from_disk()` problem can still be fixed. Your call if you want to close the bug or create a new more specific one.

**Comment by adrianeboyd** [comment, marked as answer]
You'd get the same error in spacy v2 if you've saved the component as part of a model, because the vocab isn't saved within the component in that case.

```python
# spacy v2.x
textcat.to_disk("my_textcat") # saves a vocab/ directory underneath textcat/
nlp.to_disk("my_model")       # saves one central vocab/ directory one level higher, not within textcat/
```

`from_disk` for just the `textcat/` directory would only work in the first case, not the second.

But in spacy v3 we also need information from the model config in the top-level `config.cfg`, so the `from_disk` version of adding a new pipe is no longer possible even if you saved the vocab with it. If you have an object like `TextCategorizer`, you can't add it to another pipeline anymore because of this. You have to use the `source` option from a saved model instead.

**Comment by nopper** [comment]
I see thanks again for the details. I am closing the bug then.
